### PR TITLE
x/tokenfactory:  add type files

### PR DESCRIPTION
**Pull request by sunnya97**
To be merged after #1360 

## What is the purpose of the change

Add the non-proto type files for the x/tokenfactory module

## Testing and Verifying

This change added tests and can be verified as follows:

  - *Added unit test that validates denom names and genesis files*

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? (yes / no)
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? (yes / no)
  - How is the feature or change documented? (not applicable   /   specification (`x/<module>/spec/`)  /  [Osmosis docs repo](https://github.com/osmosis-labs/docs)   /   not documented)

### CODE REVIEW

File: x/tokenfactory/types/denoms.go [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+		return "", "", sdkerrors.Wrapf(ErrInvalidDenom, "denom prefix is incorrect. Is: %s.  Should be: %s", strParts[0], ModuleDenomPrefix)
+	}
+
+	creator = strParts[1]
+	_, err = sdk.AccAddressFromBech32(creator)
+	if err != nil {
+		return "", "", sdkerrors.Wrapf(ErrInvalidDenom, "Invalid creator address (%s)", err)
+	}
+
+	// This is to for if there is a denom with a slash in its nonce. For example,
```
Comment (alexanderbez) [inline x/tokenfactory/types/denoms.go:43]:
```suggestion
	// Handle the case where a denom has a slash in its nonce. For example,
```

File: x/tokenfactory/types/denoms.go [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+)
+
+const (
+	ModuleDenomPrefix = "factory"
+)
+
+func GetTokenDenom(creator, nonce string) (string, error) {
```
Comment (alexanderbez) [inline x/tokenfactory/types/denoms.go:14]: nit: godocs for these functions :)

File: x/tokenfactory/types/errors.go [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+
+// DONTCOVER
+
+import (
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+)
+
+// x/tokenfactory module sentinel errors
+var (
+	ErrDenomExists              = sdkerrors.Register(ModuleName, 1, "denom already exists")
```
Comment (alexanderbez) [inline x/tokenfactory/types/errors.go:11]: IIRC, codes must start >= 2.

File: x/tokenfactory/types/genesis.go [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+func DefaultGenesis() *GenesisState {
+	return &GenesisState{
+		FactoryDenoms: []GenesisDenom{},
+	}
+}
+
+// Validate performs basic genesis state validation returning an error upon any
+// failure.
+func (gs GenesisState) Validate() error {
+	seenDenoms := map[string]bool{}
```
Comment (alexanderbez) [inline x/tokenfactory/types/genesis.go:23]: nit: we can be slightly more performant here and use `struct{}`, but maybe that's overkill.

File: x/tokenfactory/types/keys.go [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+// KeySeparator is used to combine parts of the keys in the store
+const KeySeparator = "|"
+
+var (
+	DenomAuthorityMetadataKey = "authoritymetadata"
+	DenomsPrefixKey           = "denoms"
+	CreatorPrefixKey          = "creator"
+	AdminPrefixKey            = "admin"
+)
+
```
Comment (alexanderbez) [inline x/tokenfactory/types/keys.go:33]: nit: godocs :P

**Review by alexanderbez — APPROVED** [review]

File: x/tokenfactory/types/codec.go [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+
+	// this line is used by starport scaffolding # 1
+	"github.com/cosmos/cosmos-sdk/types/msgservice"
+)
+
+func RegisterCodec(cdc *codec.LegacyAmino) {
+	cdc.RegisterConcrete(&MsgCreateDenom{}, "osmosis/tokenfactory/create-denom", nil)
+	cdc.RegisterConcrete(&MsgMint{}, "osmosis/tokenfactory/mint", nil)
+	cdc.RegisterConcrete(&MsgBurn{}, "osmosis/tokenfactory/burn", nil)
+	// cdc.RegisterConcrete(&MsgForceTransfer{}, "osmosis/tokenfactory/force-transfer", nil)
```
[the remark this answers is not in the document]
Comment (sunnya97) [reply x/tokenfactory/types/codec.go:16]: Yeah

File: x/tokenfactory/types/genesis.go (hunk 4 above)
Comment (sunnya97) [reply x/tokenfactory/types/genesis.go:23]: Yeah, for validate genesis, I don't think it's a big deal

File: x/tokenfactory/types/errors.go (hunk 3 above)
Comment (sunnya97) [reply x/tokenfactory/types/errors.go:11]: Oh ok, didn't know that

**Review by p0mvn — APPROVED** [review]
A bit late but LGTM!

**Pull request merged**